source/evp_hash.c

```diff
@@ -0,0 +1,298 @@
+static int hash_algorithm = -1;
+
+int hash_deprecated()
+{
+    int i;
+    SHA_CTX sha_ctx;
+    SHA256_CTX sha256_ctx;
+    SHA512_CTX sha512_ctx;
+    unsigned char md[EVP_MAX_MD_SIZE];
+
+    switch (hash_algorithm) {
+    case SHA1_ALG:
+        if (!SHA1_Init(&sha_ctx))
+            return 0;
+        for (i = 0; i < update_times; i++)
+            if (!SHA1_Update(&sha_ctx, data, sizeof(data)))
+                return 0;
+        if (!SHA1_Final(md, &sha_ctx))
+            return 0;
+        break;
+    case SHA224_ALG:
+        if (!SHA224_Init(&sha256_ctx))
+            return 0;
+        for (i = 0; i < update_times; i++)
+            if (!SHA224_Update(&sha256_ctx, data, sizeof(data)))
+                return 0;
+        if (!SHA224_Final(md, &sha256_ctx))
+            return 0;
+        break;
+    case SHA256_ALG:
+        if (!SHA256_Init(&sha256_ctx))
+            return 0;
+        for (i = 0; i < update_times; i++)
+            if (!SHA256_Update(&sha256_ctx, data, sizeof(data)))
+                return 0;
+        if (!SHA256_Final(md, &sha256_ctx))
+            return 0;
+        break;
+    case SHA384_ALG:
+        if (!SHA384_Init(&sha512_ctx))
+            return 0;
+        for (i = 0; i < update_times; i++)
+            if (!SHA384_Update(&sha512_ctx, data, sizeof(data)))
+                return 0;
+        if (!SHA384_Final(md, &sha512_ctx))
+            return 0;
+        break;
+    case SHA512_ALG:
+        if (!SHA512_Init(&sha512_ctx))
+            return 0;
+        for (i = 0; i < update_times; i++)
+            if (!SHA512_Update(&sha512_ctx, data, sizeof(data)))
+                return 0;
+        if (!SHA512_Final(md, &sha512_ctx))
+            return 0;
+        break;
+    default:
+        return 0;
+    }
+
+    return 1;
+}
+
+int hash_evp(EVP_MD_CTX *mctx, const EVP_MD *evp_md)
+{
+    int i;
+    unsigned char md[EVP_MAX_MD_SIZE];
+
+    if (!EVP_DigestInit(mctx, evp_md))
+        return 0;
+
+    for (i = 0; i < update_times; i++)
+        if (!EVP_DigestUpdate(mctx, data, sizeof(data)))
+            return 0;
+
+    if (!EVP_DigestFinal(mctx, md, NULL))
+        return 0;
+
+    return 1;
+}
+
+void do_hash(size_t num)
+{
+    OSSL_TIME time;
+    EVP_MD_CTX *mctx = NULL; 
+    const EVP_MD *evp_md = NULL;
+
+    counts[num] = 0;
+
+    if (!deprecated_api) {
+        switch (hash_algorithm) {
+        case SHA1_ALG:
+            evp_md = EVP_sha1();
+            break;
+        case SHA224_ALG:
+            evp_md = EVP_sha224();
+            break;
+        case SHA256_ALG:
+            evp_md = EVP_sha256();
+            break;
+        case SHA384_ALG:
+            evp_md = EVP_sha384();
+            break;
+        case SHA512_ALG:
+            evp_md = EVP_sha512();
+            break;
+        default:
+            err = 1;
+            return;
+        }
+
+        if ((mctx = EVP_MD_CTX_new()) == NULL)
+            return;
+    }
+
+    do {
+        if (deprecated_api) {
```

Comment:
This check should also go out from critical section. It should be decided in before the critical section which function to call as it does not change during runtime